metsim/metsim.py

```diff
@@ -209,169 +226,193 @@ def _validate_force_times(self, force_times):
+        print('force start {}'.format(pd.Timestamp(
             force_times.values[0]).to_pydatetime()))
-        logger.info('force stop {}'.format(pd.Timestamp(
+        print('force stop {}'.format(pd.Timestamp(
             force_times.values[-1]).to_pydatetime()))
 
-        logger.info('state start {}'.format(self.params['state_start']))
-        logger.info('state stop {}'.format(self.params['state_stop']))
+        print('state start {}'.format(self.params['state_start']))
+        print('state stop {}'.format(self.params['state_stop']))
 
-        logger.info('calendar {}'.format(self.params['calendar']))
+        print('calendar {}'.format(self.params['calendar']))
         if self.params['utc_offset']:
-            attrs['time'] = {'long_name': 'UTC time',
+            attrs['time'] = {'units': DEFAULT_TIME_UNITS,
+                             'long_name': 'UTC time',
                              'standard_name': 'utc_time'}
         else:
-            attrs['time'] = {'long_name': 'local time at grid location',
+            attrs['time'] = {'units': DEFAULT_TIME_UNITS,
+                             'long_name': 'local time at grid location',
                              'standard_name': 'local_time'}
 
+    @property
+    def domain(self):
+        if self._domain is None:
+            print("read_domain")
+            self._domain = io.read_domain(self.params).isel(
+                **self._domain_slice)
+        return self._domain
+
+    @property
+    def met_data(self):
+        if self._met_data is None:
+            print("read_met_data")
+            self._met_data = io.read_met_data(self.params, self.domain).isel(
+                **self._domain_slice)
+            self._met_data['elev'] = self.domain['elev']
+            self._met_data['lat'] = self.domain['lat']
+            self._validate_force_times(force_times=self._met_data['time'])
+        return self._met_data
+
+    @property
+    def state(self):
+        if self._state is None:
+            print("read_state")
+            self._state = io.read_state(self.params, self.domain).isel(
+                **self._domain_slice)
+            print("_aggregate_state")
+            self._aggregate_state()
+        return self._state
+
+    @property
+    def slices(self):
+        if not self.params['chunks']:
+            return [{d: slice(None) for d in self.domain[['mask']].dims}]
+        return chunk_domain(self.params['chunks'], self.domain[['mask']].dims)
+
+    def run(self):
+        # TODO: we should be able to pickle the Metsim and not send the params
+        # dict in this way. This feels a bit hokey.
+        print('running chunks')
+        times = self._get_output_times()
+        filename = self._get_output_filename(times)
+        self.setup_netcdf_output(filename, times)
+
+        # write_lock = _get_scheduler_lock(self.params['scheduler'], path_or_file=filename)
+        m = multiprocessing.Manager()
+        write_lock = m.Lock()
```

Comment:
we should go back to trying the `_get_scheduler_lock` approach - I'm still not clear if we need the multiprocessing lock directly.